doc/developer/design/20210401_volatile_sources.md

```diff
@@ -0,0 +1,294 @@
+    educate our users about volatile sources.
+
+  * Restrict the use of volatile sources in contexts where they could impede
+    future improvements to the system's determinism and correctness.
+
+## Goals
+
+The only goal is to **stabilize the [PubNub] and [SSE] sources**.
+
+Both of these are volatile sources with some worrying correctness properties,
+but their existence will massively improve the "wow factor" for demos of
+Materialize Cloud.
+
+[PubNub]: https://github.com/MaterializeInc/materialize/pull/6189
+[SSE]: https://github.com/MaterializeInc/materialize/pull/6208
+
+## Description
+
+### Terminology
+
+A "volatile source" is any source that cannot reliabily present the same data
+with the same timestamps across different instantiations.
+
+Examples of volatile sources:
+
+  * Any Kinesis source. Since Kinesis streams have mandatory retention periods
+    (by default 24 hours), a new instantiation of the source will not see
+    any data that has expired.
+  * A PubNub source, since new instantiations always start from the tail of the
+    stream, at least in the current implementation.
+  * A Kafka source with a non-infinite retention policy.
+
+Examples of nonvolatile sources:
+
+  * A file source, when the file is treated as append only.
+  * An S3 source, when the bucket is treated as append only.
+  * An non-upsert Kafka source, when the topic has infinite retention and no
+    compaction policy.
+  * An upsert Kafka source, when the topic has infinite retention but *may*
+    have a key compaction policy.
+
+It's important to note that volatility is a property of how a source is *used*,
+rather than fundamental to a source type. Kinesis is the rare exception. All
+other sources can be either volatile and non-volatile modes, depending on
+how the source is used.
+
+### Discussion
+
+Volatile sources impede Materialize's goal of providing users with correct,
+deterministic results. If you restart Materialize, we want to guarantee that you
+will see exactly the same data at exactly the same timestamps, and there is
+ongoing work to facilitate that. If you've used a volatile source, there is
+simply no possibility that we'll be able to provide that guarantee, because
+the upstream source may have garabage collected the data that you saw last time
+you ran Materialize.
+
+Several forthcoming initiatives depend on this determinism guarantee, however:
+
+  * Exactly-once sinks
+  * Active-active replication
+  * Dataflow transformations that consider source instances to be logically
+    equivalent
+
+My contention is that it is wholly unrealistic to expect our users to only use
+nonvolatile sources with Materialize. The streaming world runs on inconsistency
+and nondurability. To play in this space, we need to interoperate seamlessly
+with volatile sources, and then offer a large, inviting off-ramp into the
+Materialize world of determinism and consistency.
```

Comment:
These determinism problems make a lot of sense.

Fundamentally though I think the fact that materialize doesn't (in general) persist user data means that all our sources are volatile. In particular, making this distinction between e.g. files and PubNub hides the fact that you need to make sure that you don't delete your files in order for materialize to behave correctly.

I worry that marking some sources as "safer" than others hides the fact that -- again, because materialize doesn't do persistence -- the correctness properties of any system using materialized depends on the transitive configuration of all sources that materialize is reading data from.

I agree that the "volatile" bit will allow us to do a better job providing error messages with exactly-once sinks, but I worry that it will just hide subtler correctness issues, and we won't be able to provide decent error messages based solely on the bit.

We might, though be able to target and detect the actual cause of these problems. For example, to determine if two sources have (or are likely to have) the same data, we could record the first message retrieved from the first instantiation of a source, and all future instantiations could verify that they receive that first message, otherwise they're not equivalent.

Reply:
> In particular, making this distinction between e.g. files and PubNub hides the fact that you need to make sure that you don't delete your files

I don't think the distinction between files as nonvolatile and PubNub as volatile hides the fact that you can't delete your files, and it certainly shouldn't. The language in the proposal around files says

> A file source, where the file is periodically overwritten.

is a volatile source, and

> A file source, when the file is treated as append only.

is a nonvolatile source. I think we need to make it expressly clear in the docs that only certain usages of files are going to work with volatile sources (including not deleting them between restarts) and if the users can't follow those usage guidelines and declares a `NONVOLATILE` source anyway, all bets are off. If people repeatedly mess up on doing this, then I think we can start looking at autodetection of bad usage, and not before?

Reply:
> Fundamentally though I think the fact that materialize doesn't (in general) persist user data means that all our sources are volatile.

So, another way of looking at this is that when you mark a source as "nonvolatile" you are making a commitment to Materialize that your source adheres to some fiddly properties that are easy to violate. And possibly we will add additional properties to this list over time, per Brennan's note above. I think forcing users to write down the properties they claim their source supports is a huge win. The fact that we will automatically consider PubNub/SSE/Kinesis as volatile and others as nonvolatile is a temporary hack, IMO, and over time we should move towards either forcing people to specify these more explicitly.

I do think automatically looking for violations of the properties is a great idea. @umanwizard has wanted this for a while too, apparently. I added a section to the doc about that.

Reply:
Great, that all sounds reasonable. I'm not trying to block this -- which I think is a reasonable docs strategy -- but I would like us to take more ownership of ensuring that sources are behaving well and just surfacing errors when they don't.